packages/dcos-launch/extra/launch/config.py

```diff
@@ -49,375 +21,256 @@ def expand_path(path: str, relative_dir: str) -> str:
+    return launch.util.load_string(doc['ssh_private_key_filename'])
 
 
-def validate_key_helper_support(platform, key_helper):
-    # TODO DCOS-14033: [gen.internals] Source validate functions are global only
-    if key_helper == 'false':
-        return
-    assert platform in ('aws', 'azure')
+class LaunchValidator(cerberus.Validator):
+    """ Needs to use unintuitive pattern so that child validator can be created
+    for validated the nested dcos_config
+    """
+    def __init__(self, *args, **kwargs):
+        super(LaunchValidator, self).__init__(*args, **kwargs)
+        if 'config_dir' in kwargs:
+            self.config_dir = kwargs['config_dir']
 
+    def _normalize_coerce_expand_local_path(self, value):
+        return expand_path(value, self.config_dir)
 
-def calculate_ssh_user(os_name, platform):
-    if platform == 'aws':
-        return test_util.aws.OS_SSH_INFO[os_name].user
-    else:
-        raise Exception('Cannot yet calculate user for {} platform'.format(platform))
 
+def expand_error_dict(errors: dict) -> str:
+    message = ''
+    for key, errors in errors.items():
+        sub_message = 'Field: {}, Errors: '.format(key)
+        for e in errors:
+            if isinstance(e, dict):
+                sub_message += expand_error_dict(e)
+            else:
+                sub_message += e
+            sub_message += '\n'
+        message += sub_message
+    return message
 
-def validate_key_helper_parameters(template_parameters, provider, key_helper):
-    # TODO DCOS-14033: [gen.internals] Source validate functions are global only
-    if key_helper == 'false':
-        return
-    if provider == 'aws':
-        assert 'KeyName' not in yaml.load(template_parameters), 'key_helper will '\
-            'automatically calculate and inject KeyName; do not set this parameter'
-    if provider == 'azure':
-        assert 'sshRSAPublicKey' not in yaml.load(template_parameters), 'key_helper will '\
-            'automatically calculate and inject sshRSAPublicKey; do not set this parameter'
 
+def raise_errors(validator: LaunchValidator):
+    message = expand_error_dict(validator.errors)
+    raise launch.util.LauncherError('ValidationError', message)
 
-def calculate_instance_count(num_masters, num_private_agents, num_public_agents):
-    return str(1 + int(num_masters) + int(num_private_agents) + int(num_public_agents))
 
+def get_validated_config(config_path: str) -> dict:
+    """ Returns validated a finalized argument dictionary for dcos-launch
+    Given the huge range of configuration space provided by this configuration
+    file, it must be processed in three steps (common, provider-specifc,
+    platform-specific)
+    """
+    config = load_config(config_path)
+    config_dir = os.path.dirname(config_path)
+    # validate against the fields common to all configs
+    basic_validator = LaunchValidator(COMMON_SCHEMA, config_dir=config_dir, allow_unknown=True)
+    if not basic_validator.validate(config):
```

Comment:
I liked keeping the code this way because basic_validator.validate() is a state-mutating function and so I wanted to keep all the mutation at one level. Also, I really dislike functions that do too much and to put this into a single function would be something like:
```def validate_and_raise_if_errors(validator: LaunchValidator, config:dict) -> None```
which I thought just confused thing more.
Thoughts?